spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala

```diff
@@ -547,25 +574,273 @@ object Checkpoints extends DeltaLogging {
+      // In case of V2 Checkpoints, zero row count is possible.
       logWarning(DeltaErrors.EmptyCheckpointErrorMessage)
     }
+
+    // If we don't parallelize, we use None for backwards compatibility
+    val checkpointParts = if (legacyMultiPartCheckpoint) Some(numParts) else None
+
     LastCheckpointInfo(
       version = snapshot.version,
-      size = checkpointRowCount.value,
-      parts = numPartsOption,
-      sizeInBytes = Some(checkpointSizeInBytes),
+      size = overallNumCheckpointActions,
+      parts = checkpointParts,
+      sizeInBytes = Some(overallCheckpointSizeInBytes),
       numOfAddFiles = Some(snapshot.numOfFiles),
-      checkpointSchema = checkpointSchemaToWriteInLastCheckpointFile(spark, schema)
+      v2Checkpoint = v2Checkpoint,
+      checkpointSchema = checkpointSchemaToWriteInLastCheckpoint
     )
   }
 
+  /**
+   * Generate a tuple of the file to write the checkpoint and where it may later need
+   * to be copied. Should be used within a task, so that task or stage retries don't
+   * create the same files.
+   */
+  def getCheckpointWritePath(
+      conf: Configuration,
+      logPath: String,
+      version: Long,
+      numParts: Int,
+      part: Int,
+      useRename: Boolean,
+      v2CheckpointEnabled: Boolean): (Path, Path) = {
+    def getCheckpointWritePath(path: Path): Path = {
+      if (useRename) {
+        val tempPath =
+          new Path(path.getParent, s".${path.getName}.${UUID.randomUUID}.tmp")
+        DeltaFileOperations.registerTempFileDeletionTaskFailureListener(conf, tempPath)
+        tempPath
+      } else {
+        path
+      }
+    }
+    val logPathObj = new Path(logPath)
+    val destinationName: Path = if (v2CheckpointEnabled) {
+      newV2CheckpointSidecarFile(logPathObj, version, numParts, part + 1)
+    } else {
+      if (numParts > 1) {
+        assert(part < numParts, s"Asked to create part: $part of max $numParts in checkpoint.")
+        checkpointFileWithParts(logPathObj, version, numParts)(part)
+      } else {
+        checkpointFileSingular(logPathObj, version)
+      }
+    }
+
+    getCheckpointWritePath(destinationName) -> destinationName
+  }
+
+  /**
+   * Writes a top-level V2 Checkpoint file which may point to multiple
+   * sidecar files.
+   *
+   * @return A tuple containing
+   *          1. [[FileStatus]] of the newly created top-level V2Checkpoint.
+   *          2. The sequence of actions that were written to the top-level file.
+   *          3. An instance of the LastCheckpointV2 containing V2-checkpoint related
+   *           metadata which can later be written to LAST_CHECKPOINT
+   *          4. Schema of the newly written top-level file (only for parquet files)
+   */
+  protected[delta] def writeTopLevelV2Checkpoint(
+      v2CheckpointFormatOpt: Option[V2Checkpoint.Format],
+      sidecarCheckpointFiles: Array[SerializableFileStatus],
+      spark: SparkSession,
+      schema: StructType,
```

Comment:
Some of the arguments are non-trivial (e.g. schema), could we document the parameters in the method doc?